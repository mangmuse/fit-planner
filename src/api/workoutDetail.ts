import { BASE_URL } from "@/constants";
import { db } from "@/lib/db";
import { LocalWorkoutDetail } from "@/types/models";

export type SyncWorkoutDetailsPayload = {
  mappedUnsynced: (Omit<LocalWorkoutDetail, "workoutId"> & {
    workoutId: string;
  })[];
};

export const syncToServerWorkoutDetails = async () => {
  // 1. 디테일 다가져와
  const all = await db.workoutDetails.toArray();
  // 2. isSynced false인거 거르고
  const unsynced = all.filter((detail) => !detail.isSynced);
  // 3. 각각 매핑 -> exerciseId로 exercise 가져와서 serverId로 매핑,
  // 4. workoutId로 workout 가져와서 serverId로 매핑,
  const mappedUnsynced = await Promise.all(
    unsynced.map(async (detail) => {
      const exercise = await db.exercises
        .where("id")
        .equals(detail.exerciseId)
        .first();
      const workout = await db.workouts
        .where("id")
        .equals(detail.workoutId)
        .first();
      return {
        ...detail,
        exerciseId: exercise?.serverId,
        workoutId: workout?.serverId,
      };
    })
  );

  console.log(mappedUnsynced);
  // 5. route로 보내서 detail에 serverId가 있으면 업데이트 없으면 create

  const res = await fetch(`${BASE_URL}/api/workout/detail/sync`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ mappedUnsynced }),
  });

  if (!res.ok) throw new Error("WorkoutDetails 동기화에 실패했습니다");

  const data = await res.json();
  console.log(data);

  if (data.updated) {
    for (const updated of data.updated) {
      const exercise = await db.exercises
        .where("serverId")
        .equals(updated.exerciseId)
        .first();
      const workout = await db.workouts
        .where("serverId")
        .equals(updated.workoutId)
        .first();
      await db.workoutDetails.update(updated.localId, {
        serverId: updated.serverId,
        isSynced: true,
        exerciseId: exercise?.id,
        workoutId: workout?.id,
      });
    }
  }
};
