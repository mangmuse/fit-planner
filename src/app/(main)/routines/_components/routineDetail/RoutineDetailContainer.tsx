"use client";
import SessionPlaceholder from "@/app/(main)/_shared/session/SessionPlaceholder";
import { TEMP_ROUTINE_ID } from "@/app/(main)/routines/constants";

import { routineDetailService } from "@/lib/di";
import { useParams } from "next/navigation";
import { useEffect, useState } from "react";

type RoutineDetailContainerProps = {
  prevData?: [];
  reload?: () => Promise<void>;
};

const RoutineDetailContainer = ({
  prevData,
  reload,
}: RoutineDetailContainerProps) => {
  const [isLoading, setIsLoading] = useState<boolean>(true);

  const { routineId } = useParams();
  const getDetails = async () => {
    const id = routineId ?? TEMP_ROUTINE_ID;
    const good = await routineDetailService.getLocalRoutineDetails(Number(id));
  };

  useEffect(() => {
    // reload();
    getDetails();
  }, []);
  return (
    <>{prevData ? <div>hello</div> : <SessionPlaceholder type="ROUTINE" />}</>
  );
};

export default RoutineDetailContainer;
