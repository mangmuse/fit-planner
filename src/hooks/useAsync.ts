import { useState, useEffect, useCallback } from "react";

export const useAsync = <T>(
  asyncFunction: () => Promise<T>,
  dependencies: unknown[] = []
) => {
  const [data, setData] = useState<T | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [error, setError] = useState<Error | null>(null);

  const execute = useCallback(async () => {
    setIsLoading(true);
    setError(null);
    try {
      const result = await asyncFunction();
      setData(result);
    } catch (err) {
      const error = err as Error;
      if (process.env.NODE_ENV === "development") {
        console.error("[useAsync] error:", error);
      }
      setError(error);
    } finally {
      setIsLoading(false);
    }
  }, [asyncFunction]);

  useEffect(() => {
    execute();
  }, dependencies);

  return { data, isLoading, error, execute };
};
