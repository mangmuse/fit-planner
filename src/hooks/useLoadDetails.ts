import {
  isWorkoutDetail,
  isWorkoutDetails,
} from "@/app/(main)/workout/_utils/checkIsWorkoutDetails";
import { getGroupedDetails } from "@/app/(main)/workout/_utils/getGroupedDetails";
import {
  LocalRoutineDetail,
  LocalWorkout,
  LocalWorkoutDetail,
  Saved,
} from "@/types/models";
import { useEffect, useMemo, useState, useCallback } from "react";
import {
  routineDetailService,
  workoutDetailService,
  workoutService,
} from "@/lib/di";

type UseLoadDetailsProps = {
  type: "RECORD" | "ROUTINE";
  userId: string;
  date?: string;
  routineId?: number;
};

export type SessionGroup = {
  exerciseOrder: number;
  details: Saved<LocalWorkoutDetail>[] | Saved<LocalRoutineDetail>[];
};

const useLoadDetails = ({
  type,
  userId,
  date,
  routineId,
}: UseLoadDetailsProps) => {
  const [workout, setWorkout] = useState<LocalWorkout | null>(null);
  const [error, setError] = useState<string | null>(null);

  const [isInitialLoading, setIsInitialLoading] = useState<boolean>(true);
  const [allDetails, setAllDetails] = useState<
    Saved<LocalWorkoutDetail>[] | Saved<LocalRoutineDetail>[]
  >([]);
  const [workoutGroups, setWorkoutGroups] = useState<SessionGroup[]>([]);

  const loadLocalDetails = useCallback(async () => {
    try {
      setIsInitialLoading(true);
      if (type === "RECORD") {
        if (!userId || !date) return;
        const details = await workoutDetailService.getLocalWorkoutDetails(
          userId,
          date
        );
        setAllDetails(details);
        const adjustedGroups = getGroupedDetails(details);

        setWorkoutGroups(adjustedGroups);
      } else if (type === "ROUTINE") {
        if (!userId || !routineId) return;
        const details =
          await routineDetailService.getLocalRoutineDetails(routineId);
        setAllDetails(details);
        const adjustedGroups = getGroupedDetails(details);
        setWorkoutGroups(adjustedGroups);
      }
    } catch (e) {
      console.error(e);
      setError("운동 세부 정보를 불러오는데 실패했습니다");
    } finally {
      setIsInitialLoading(false);
    }
  }, [type, userId, date, routineId]);

  const syncWorkoutStatus = async () => {
    try {
      if (!date || !userId) return;

      let currentWorkout = workout;
      if (!currentWorkout) {
        const fetchedWorkout = await workoutService.getWorkoutByUserIdAndDate(
          userId,
          date
        );
        currentWorkout = fetchedWorkout || null;
        if (!currentWorkout) {
          // 최초 렌더링이나 workout이 아직 생성되지 않은 경우
          return;
        }
        setWorkout(currentWorkout);
      }
      if (!currentWorkout?.id || currentWorkout.status === "COMPLETED") return;
      const newStatus = workoutGroups.length === 0 ? "EMPTY" : "PLANNED";
      await workoutService.updateLocalWorkout({
        ...currentWorkout,
        status: newStatus,
      });
    } catch (e) {
      console.error("[useLoadDetails] Error", e);
      setError("운동 상태를 동기화하는데 실패했습니다");
    }
  };

  useEffect(() => {
    (async () => {
      await loadLocalDetails();
    })();
  }, [type, userId, date, routineId]);

  useEffect(() => {
    if (type === "RECORD" && date) {
      syncWorkoutStatus();
    }
  }, [workoutGroups]);

  const updateDetailInGroups = useCallback(
    (updatedDetail: Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>) => {
      setWorkoutGroups((prevGroups) =>
        prevGroups.map((group) => {
          const hasDetail = group.details.some(
            (d) => d.id === updatedDetail.id
          );
          if (!hasDetail) return group; // 참조 유지

          return {
            ...group,
            details: group.details.map((d) =>
              d.id === updatedDetail.id ? updatedDetail : d
            ),
          };
        })
      );

      setAllDetails((prev) =>
        prev.map((d) => (d.id === updatedDetail.id ? updatedDetail : d))
      );
    },
    []
  );

  const addDetailToGroup = useCallback(
    (
      newDetail: Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>,
      lastDetail: Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>
    ) => {
      setWorkoutGroups((prevGroups) =>
        prevGroups.map((group) => {
          const hasLastDetail = group.details.some(
            (d) => d.id === lastDetail.id
          );
          if (!hasLastDetail) return group;

          if (isWorkoutDetails(group.details) && isWorkoutDetail(newDetail)) {
            const lastIndex = group.details.findIndex(
              (d) => d.id === lastDetail.id
            );
            const newDetails = [...group.details];
            newDetails.splice(lastIndex + 1, 0, newDetail);

            return {
              ...group,
              details: newDetails,
            };
          } else if (
            !isWorkoutDetails(group.details) &&
            !isWorkoutDetail(newDetail)
          ) {
            const lastIndex = group.details.findIndex(
              (d) => d.id === lastDetail.id
            );
            const newDetails = [...group.details];
            newDetails.splice(lastIndex + 1, 0, newDetail);

            return {
              ...group,
              details: newDetails,
            };
          }

          return group;
        })
      );

      setAllDetails((prev) => {
        if (isWorkoutDetails(prev) && isWorkoutDetail(newDetail)) {
          const lastIndex = prev.findIndex((d) => d.id === lastDetail.id);
          const newAll = [...prev];
          newAll.splice(lastIndex + 1, 0, newDetail);
          return newAll;
        } else if (!isWorkoutDetails(prev) && !isWorkoutDetail(newDetail)) {
          const lastIndex = prev.findIndex((d) => d.id === lastDetail.id);
          const newAll = [...prev];
          newAll.splice(lastIndex + 1, 0, newDetail);
          return newAll;
        }
        return prev;
      });
    },
    []
  );

  const removeDetailFromGroup = useCallback((detailId: number) => {
    setWorkoutGroups((prevGroups) => {
      let deletedOrder: number | null = null;

      const filteredGroups = prevGroups
        .map((group) => {
          const hasDetail = group.details.some((d) => d.id === detailId);
          if (!hasDetail) return group;

          const newDetails = group.details.filter((d) => d.id !== detailId);

          if (newDetails.length === 0) {
            deletedOrder = group.exerciseOrder;
            return null;
          }

          return {
            ...group,
            details: newDetails,
          };
        })
        .filter(Boolean) as SessionGroup[];

      if (deletedOrder !== null) {
        return filteredGroups.map((group) => ({
          ...group,
          exerciseOrder:
            group.exerciseOrder > deletedOrder!
              ? group.exerciseOrder - 1
              : group.exerciseOrder,
        }));
      }

      return filteredGroups;
    });

    setAllDetails((prev) => {
      if (isWorkoutDetails(prev)) {
        return prev.filter((d) => d.id !== detailId);
      } else {
        return prev.filter((d) => d.id !== detailId);
      }
    });
  }, []);

  return {
    error,
    isLoading: isInitialLoading,
    allDetails,
    setAllDetails,
    workout,
    workoutGroups,
    reload: loadLocalDetails,
    setWorkout,
    updateDetailInGroups,
    addDetailToGroup,
    removeDetailFromGroup,
  };
};

export default useLoadDetails;
