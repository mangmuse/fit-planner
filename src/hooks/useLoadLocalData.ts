type UseLoadLocalDataProps = {
  getDataCallback: () => void;
  setStateCallback: () => void;
};

const useLoadLocalData = ({
  getDataCallback,
  setStateCallback,
}: UseLoadLocalDataProps) => {};

export default useLoadLocalData;
