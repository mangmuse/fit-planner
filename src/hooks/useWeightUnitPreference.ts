import { useState, useEffect } from "react";

const WEIGHT_UNIT_KEY = "preferred_weight_unit";

export function useWeightUnitPreference() {
  const [unit, setUnit] = useState<"kg" | "lbs">(() => {
    if (typeof window === "undefined") return "kg";
    return (localStorage.getItem(WEIGHT_UNIT_KEY) as "kg" | "lbs") || "kg";
  });

  const setPreferredUnit = (newUnit: "kg" | "lbs") => {
    setUnit(newUnit);
    localStorage.setItem(WEIGHT_UNIT_KEY, newUnit);
  };

  useEffect(() => {
    const handleStorageChange = (e: StorageEvent) => {
      if (e.key === WEIGHT_UNIT_KEY && e.newValue) {
        setUnit(e.newValue as "kg" | "lbs");
      }
    };

    window.addEventListener("storage", handleStorageChange);
    return () => window.removeEventListener("storage", handleStorageChange);
  }, []);

  return [unit, setPreferredUnit] as const;
}

export function getPreferredWeightUnit(): "kg" | "lbs" {
  if (typeof window === "undefined") return "kg";
  return (localStorage.getItem(WEIGHT_UNIT_KEY) as "kg" | "lbs") || "kg";
}
