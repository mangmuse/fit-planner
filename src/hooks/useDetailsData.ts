import { getGroupedDetails } from "@/app/(main)/workout/_utils/getGroupedDetails";
import { SessionGroup } from "@/hooks/useLoadDetails";
import {
  routineDetailService,
  workoutDetailService,
  workoutService,
} from "@/lib/di";
import {
  LocalRoutineDetail,
  LocalWorkout,
  LocalWorkoutDetail,
  Saved,
} from "@/types/models";
import { SessionDetailType } from "@/types/services";
import { useCallback, useEffect, useState } from "react";

export const useDetailsData = (
  type: SessionDetailType,
  userId: string,
  date?: string,
  routineId?: number
) => {
  const [workout, setWorkout] = useState<Saved<LocalWorkout> | null>(null);
  const [error, setError] = useState<string | null>(null);

  const [isInitialLoading, setIsInitialLoading] = useState<boolean>(true);
  const [workoutGroups, setWorkoutGroups] = useState<SessionGroup[]>([]);

  const loadLocalDetails = useCallback(async () => {
    try {
      setIsInitialLoading(true);
      if (type === "RECORD") {
        if (!userId || !date) return;
        const details = await workoutDetailService.getLocalWorkoutDetails(
          userId,
          date
        );
        const adjustedGroups = getGroupedDetails(details);
        setWorkoutGroups(adjustedGroups);
      } else if (type === "ROUTINE") {
        if (!userId || !routineId) return;
        const details =
          await routineDetailService.getLocalRoutineDetails(routineId);
        const adjustedGroups = getGroupedDetails(details);
        setWorkoutGroups(adjustedGroups);
      }
    } catch (e) {
      console.error(e);
      setError("운동 세부 정보를 불러오는데 실패했습니다");
    } finally {
      setIsInitialLoading(false);
    }
  }, [type, userId, date, routineId]);

  const syncWorkoutStatus = async () => {
    try {
      if (!date || !userId) return;

      let currentWorkout = workout;
      if (!currentWorkout) {
        const fetchedWorkout = await workoutService.getWorkoutByUserIdAndDate(
          userId,
          date
        );
        currentWorkout = fetchedWorkout || null;
        if (!currentWorkout) {
          // 최초 렌더링이나 workout이 아직 생성되지 않은 경우
          return;
        }
        setWorkout(currentWorkout);
      }
      if (!currentWorkout?.id || currentWorkout.status === "COMPLETED") return;
      const newStatus = workoutGroups.length === 0 ? "EMPTY" : "PLANNED";
      await workoutService.updateLocalWorkout({
        ...currentWorkout,
        status: newStatus,
      });
    } catch (e) {
      console.error("[useLoadDetails] Error", e);
      setError("운동 상태를 동기화하는데 실패했습니다");
    }
  };

  useEffect(() => {
    (async () => {
      await loadLocalDetails();
    })();
  }, [type, userId, date, routineId]);

  useEffect(() => {
    if (type === "RECORD" && date) {
      syncWorkoutStatus();
    }
  }, [type, userId, date, routineId]);

  return {
    data: {
      workout,
      workoutGroups,
    },
    setData: {
      setWorkoutGroups,
      setWorkout,
    },
    isLoading: isInitialLoading,
    error,
    reload: loadLocalDetails,
  };
};
