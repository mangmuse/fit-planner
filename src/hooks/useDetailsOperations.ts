import {
  isWorkoutDetail,
  isWorkoutDetails,
} from "@/app/(main)/workout/_utils/checkIsWorkoutDetails";
import { SessionGroup } from "@/hooks/useLoadDetails";
import { LocalRoutineDetail, LocalWorkoutDetail, Saved } from "@/types/models";
import { useCallback, Dispatch, SetStateAction } from "react";

type UseDetailsOperationsProps = {
  workoutGroups: SessionGroup[];
  setWorkoutGroups: Dispatch<SetStateAction<SessionGroup[]>>;
};

const insertDetailAfterLastDetail = <
  T extends Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>,
>(
  details: T[],
  lastDetail: T,
  newDetail: T
): T[] => {
  const lastIndex = details.findIndex((d) => d.id === lastDetail.id);
  const newDetails = [...details];
  newDetails.splice(lastIndex + 1, 0, newDetail);
  return newDetails;
};

export const useDetailsOperations = ({
  workoutGroups,
  setWorkoutGroups,
}: UseDetailsOperationsProps) => {
  const updateDetailInGroups = useCallback(
    (updatedDetail: Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>) => {
      setWorkoutGroups((prevGroups: SessionGroup[]) =>
        prevGroups.map((group) => {
          const hasDetail = group.details.some(
            (d) => d.id === updatedDetail.id
          );
          if (!hasDetail) return group; // 참조 유지

          return {
            ...group,
            details: group.details.map((d) =>
              d.id === updatedDetail.id ? updatedDetail : d
            ),
          };
        })
      );
    },
    [setWorkoutGroups]
  );

  const addDetailToGroup = useCallback(
    (
      newDetail: Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>,
      lastDetail: Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>
    ) => {
      setWorkoutGroups((prevGroups: SessionGroup[]) =>
        prevGroups.map((group) => {
          const hasLastDetail = group.details.some(
            (d) => d.id === lastDetail.id
          );
          if (!hasLastDetail) return group;

          const isWorkoutGroup = isWorkoutDetails(group.details);
          const isWorkoutNewDetail = isWorkoutDetail(newDetail);

          if (isWorkoutGroup === isWorkoutNewDetail) {
            if (isWorkoutGroup) {
              const newDetails = insertDetailAfterLastDetail(
                group.details as Saved<LocalWorkoutDetail>[],
                lastDetail as Saved<LocalWorkoutDetail>,
                newDetail as Saved<LocalWorkoutDetail>
              );
              return {
                ...group,
                details: newDetails,
              };
            } else {
              const newDetails = insertDetailAfterLastDetail(
                group.details as Saved<LocalRoutineDetail>[],
                lastDetail as Saved<LocalRoutineDetail>,
                newDetail as Saved<LocalRoutineDetail>
              );
              return {
                ...group,
                details: newDetails,
              };
            }
          }

          return group;
        })
      );
    },
    [setWorkoutGroups]
  );

  const removeDetailFromGroup = useCallback(
    (detailId: number) => {
      setWorkoutGroups((prevGroups: SessionGroup[]) => {
        let deletedOrder: number | null = null;

        const filteredGroups = prevGroups
          .map((group) => {
            const hasDetail = group.details.some((d) => d.id === detailId);
            if (!hasDetail) return group;

            const newDetails = group.details.filter((d) => d.id !== detailId);

            if (newDetails.length === 0) {
              deletedOrder = group.exerciseOrder;
              return null;
            }

            return {
              ...group,
              details: newDetails,
            };
          })
          .filter(Boolean) as SessionGroup[];

        if (deletedOrder !== null) {
          return filteredGroups.map((group) => ({
            ...group,
            exerciseOrder:
              group.exerciseOrder > deletedOrder!
                ? group.exerciseOrder - 1
                : group.exerciseOrder,
          }));
        }

        return filteredGroups;
      });
    },
    [setWorkoutGroups]
  );

  const updateMultipleDetailsInGroups = useCallback(
    (
      updatedDetails: (Saved<LocalWorkoutDetail> | Saved<LocalRoutineDetail>)[]
    ) => {
      const updatedDetailsMap = new Map(
        updatedDetails.map((detail) => [detail.id, detail])
      );

      setWorkoutGroups((prevGroups: SessionGroup[]) =>
        prevGroups.map((group) => {
          const hasChanges = group.details.some((detail) =>
            updatedDetailsMap.has(detail.id)
          );

          if (!hasChanges) return group; // 참조 유지

          return {
            ...group,
            details: group.details.map((detail) => {
              const updatedDetail = updatedDetailsMap.get(detail.id);
              return updatedDetail || detail;
            }),
          };
        })
      );
    },
    [setWorkoutGroups]
  );

  const removeMultipleDetailsInGroup = useCallback(
    (details: Saved<LocalWorkoutDetail>[] | Saved<LocalRoutineDetail>[]) => {
      setWorkoutGroups((prevGroups: SessionGroup[]) =>
        prevGroups
          .filter(
            (group) =>
              !details.some((d) => group.details.some((g) => g.id === d.id))
          )
          .map((group, idx) => ({
            ...group,
            exerciseOrder: idx + 1,
          }))
      );
    },
    [setWorkoutGroups]
  );

  return {
    removeDetailFromGroup,
    updateDetailInGroups,
    addDetailToGroup,
    updateMultipleDetailsInGroups,
    removeMultipleDetailsInGroup,
  };
};
