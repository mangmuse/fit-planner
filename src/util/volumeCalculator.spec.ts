import { mockWorkoutDetail } from "@/__mocks__/workoutDetail.mock";
import {
  calculateTotalVolume,
  calculateVolumeFromDetails,
} from "./volumeCalculator";
import { mockRoutineDetail } from "@/__mocks__/routineDetail.mock";
import { SessionGroup } from "@/hooks/useLoadDetails";

const mockRD = mockRoutineDetail.past;
const mockWD = mockWorkoutDetail.past;

describe("calculateVolumeFromDetails", () => {
  it("빈 배열인경우 0을 반환해야한다", () => {
    const details = [];
    const result = calculateVolumeFromDetails(details);
    expect(result).toBe(0);
  });
  describe("workoutDetail", () => {
    it("올바르게 볼륨을 계산한다", () => {
      const details = [
        { ...mockWD, weight: 100, reps: 10 },
        { ...mockWD, weight: 200, reps: 20 },
        { ...mockWD, weight: 300, reps: 30 },
      ];

      const details2 = [
        { ...mockWD, weight: 0, reps: 100 },
        { ...mockWD, weight: 100, reps: 0 },
      ];

      const result = calculateVolumeFromDetails(details);
      const result2 = calculateVolumeFromDetails(details2);

      expect(result).toBe(14000);
      expect(result).toMatchSnapshot();

      expect(result2).toBe(0);
      expect(result2).toMatchSnapshot();
    });
  });

  describe("routineDetail", () => {
    it("올바르게 볼륨을 계산한다", () => {
      const details = [
        { ...mockRD, weight: 100, reps: 10 },
        { ...mockRD, weight: 200, reps: 20 },
        { ...mockRD, weight: 300, reps: 30 },
      ];

      const details2 = [
        { ...mockRD, weight: 0, reps: 100 },
        { ...mockRD, weight: 100, reps: 0 },
      ];

      const result = calculateVolumeFromDetails(details);
      const result2 = calculateVolumeFromDetails(details2);

      expect(result).toBe(14000);
      expect(result).toMatchSnapshot();

      expect(result2).toBe(0);
      expect(result2).toMatchSnapshot();
    });
  });
});

describe("calculateTotalVolume", () => {
  const mockWorkoutGroup1: SessionGroup = {
    exerciseOrder: 1,
    details: [{ ...mockWD, weight: 100, reps: 10 }],
  };
  const mockWorkoutGroup2: SessionGroup = {
    exerciseOrder: 2,
    details: [
      { ...mockWD, weight: 200, reps: 20 },
      { ...mockWD, weight: 300, reps: 30 },
    ],
  };

  it("빈 배열인 경우 0을 반환해야한다", () => {
    const workoutGroups = [];
    const result = calculateTotalVolume(workoutGroups);
    expect(result).toBe(0);
  });

  const workoutGroups: SessionGroup[] = [mockWorkoutGroup1, mockWorkoutGroup2];

  it("올바르게 볼륨을 계산한다", () => {
    const result = calculateTotalVolume(workoutGroups);
    expect(result).toBe(14000);
    expect(result).toMatchSnapshot();
  });
});
