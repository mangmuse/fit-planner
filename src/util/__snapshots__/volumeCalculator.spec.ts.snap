// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateTotalVolume 올바르게 볼륨을 계산한다 1`] = `14000`;

exports[`calculateVolumeFromDetails routineDetail 올바르게 볼륨을 계산한다 1`] = `14000`;

exports[`calculateVolumeFromDetails routineDetail 올바르게 볼륨을 계산한다 2`] = `0`;

exports[`calculateVolumeFromDetails workoutDetail 올바르게 볼륨을 계산한다 1`] = `14000`;

exports[`calculateVolumeFromDetails workoutDetail 올바르게 볼륨을 계산한다 2`] = `0`;
